kafka/source/pkg/reconciler/kafkasource.go

```diff
@@ -206,6 +219,188 @@ func checkResourcesStatus(src *v1alpha1.KafkaSource) error {
 	return nil
 }
 
+func (r *Reconciler) scaleKafkaSource(ctx context.Context, ra *v1.Deployment, src *v1alpha1.KafkaSource) {
+	r.readScalingAnnotations(ctx, src)
+	// no scaling annotatins so no scaling work needed
+	if r.minReplicaCount == nil && r.maxReplicaCount == nil {
```

Comment:
one thing that I did for our GCP sources, was to have two separate controllers (see https://github.com/google/knative-gcp/blob/26efd4cf70d23a7d0fde2231d941608ebf8da98b/cmd/controller/main.go#L43) and filter the objects based on annotations. If the keda scaling annotation was on, then only the keda one gets called (see https://github.com/google/knative-gcp/blob/26efd4cf70d23a7d0fde2231d941608ebf8da98b/pkg/reconciler/pubsub/pullsubscription/keda/controller.go#L98), otherwise the "static" other...
Most of the code can be moved to a base.Reconciler and just the data plane pieces should be different. 
With this, you have a better separation of things, and don't have to do this if statements. If in the near future, we introduce another scaling mechanism, then it's easier...
I already upstreamed to pkg those filtering functions, so you can leverage it as well.
Just a thought...

Reply:
Another benefit is that old code will remain "untouched", there will be less friction with this addition... even if folks don't want keda scaling... just comment a line in the cmd main

Reply:
How do you avoid code duplicaion? There are two controllers that share lot of functionality?

Reply:
using a base reconciler. For example, see: https://github.com/google/knative-gcp/blob/26efd4cf70d23a7d0fde2231d941608ebf8da98b/pkg/reconciler/pubsub/pullsubscription/keda/pullsubscription.go#L58